docs/man/xenwatchdogd.8.pod
=head1 NAME

xenwatchdogd - Xen hypercall based watchdog daemon

=head1 SYNOPSIS

B<xenwatchdogd> [ I<OPTIONS> ] <I<TIMEOUT>> [ <I<SLEEP>> ]

=head1 DESCRIPTION

B<xenwatchdogd> arms the Xen watchdog timer to I<TIMEOUT> every I<SLEEP>
seconds. If the xenwatchdogd process dies or is delayed for more than
I<TIMEOUT> seconds, then Xen will take the B<on_watchdog> action
specified in the domain configuration (see xl.cfg(5)). If B<xenwatchdogd>
is running in dom0, the whole system will reboot.

=head1 OPTIONS

=over 4

=item B<-h>, B<--help>

Display a help message.

=item B<-F>, B<--foreground>

Run in the foreground. The default behaviour is to daemonize.

=item B<-x>, B<--safe-exit>

Disable watchdog on orderly exit. The default behaviour is to arm the
watchdog to 300 seconds to allow time for the domain to shutdown.  See 
also the B<SIGNALS> section.

=item B<timeout>

The number of seconds to arm the Xen watchdog timer. This must be set to
a minimum of two.

=item B<sleep>

The number of seconds to sleep in between calls to arm the Xen watchdog
timer. This must be at least one second, and less than the I<timeout>
value. If not specified, it defaults to half the I<timeout> value.

=back

=head1 SIGNALS

B<SIGUSR1> Will cause the program to disarm the watchdog timer and exit,
regardless of whether the B<--safe-exit> option was passed.

=head1 AUTHOR

Citrix Ltd and other contributors.


CHANGELOG.md
# Changelog

Notable changes to Xen will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)

## [4.19.0 UNRELEASED](https://xenbits.xenproject.org/gitweb/?p=xen.git;a=shortlog;h=staging) - TBD

### Changed
 - Changed flexible array definitions in public I/O interface headers to not
   use "1" as the number of array elements.
 - The minimum supported OCaml toolchain version is now 4.05
 - On x86:
   - HVM PIRQs are disabled by default.
   - Reduce IOMMU setup time for hardware domain.
 - xl/libxl configures vkb=[] for HVM domains with priority over vkb_device.
 - Increase the maximum number of CPUs Xen can be built for from 4095 to
   16383.
 - When building with Systemd support (./configure --enable-systemd), remove
   libsystemd as a build dependency.  Systemd Notify support is retained, now
   using a standalone library implementation.

### Added
 - On x86:
   - Introduce a new x2APIC driver that uses Cluster Logical addressing mode
     for IPIs and Physical addressing mode for external interrupts.
 - Add a new 9pfs backend running as a daemon in dom0. First user is
   Xenstore-stubdom now being able to support full Xenstore trace capability.

### Removed
- caml-stubdom.  It hasn't built since 2014, was pinned to Ocaml 4.02, and has
  been superseded by the MirageOS/SOLO5 projects.
- /usr/bin/pygrub symlink.  This was deprecated in Xen 4.2 (2012) but left for
  compatibility reasons.  VMs configured with bootloader="/usr/bin/pygrub"
  should be updated to just bootloader="pygrub".
- The Xen gdbstub on x86.

## [4.18.0](https://xenbits.xenproject.org/gitweb/?p=xen.git;a=shortlog;h=RELEASE-4.18.0) - 2023-11-16

### Changed
 - Repurpose command line gnttab_max_{maptrack_,}frames options so they don't
   cap toolstack provided values.
 - Ignore VCPUOP_set_singleshot_timer's VCPU_SSHOTTMR_future flag. The only
   known user doesn't use it properly, leading to in-guest breakage.
 - The "dom0" option is now supported on Arm and "sve=" sub-option can be used
   to enable dom0 guest to use SVE/SVE2 instructions.
 - Physical CPU Hotplug downgraded to Experimental and renamed "ACPI CPU
   Hotplug" for clarity

### Added
 - On x86:
   - On all Intel systems, MSR_ARCH_CAPS is now visible in guests, and
     controllable from the VM's config file.  For CPUs from ~2019 onwards,
     this allows guest kernels to see details about hardware fixes for
     speculative mitigations.  (Backported as XSA-435 to older releases).
   - xl/libxl can customize SMBIOS strings for HVM guests.
   - Support for enforcing system-wide operation in Data Operand Independent
     Timing Mode.
   - Add Intel Hardware P-States (HWP) cpufreq driver.
   - Support for features new in AMD Genoa CPUs:
     - CPUID_USER_DIS (CPUID Faulting) used by Xen to control PV guest's view
       of CPUID data.
   - Support for features new in Intel Sapphire Rapids CPUs:
     - PKS (Protection Key Supervisor) available to HVM/PVH guests.
     - VM-Notify used by Xen to mitigate certain micro-architectural pipeline
       livelocks, instead of crashing the entire server.
     - Bus-lock detection, used by Xen to mitigate (by rate-limiting) the
       system wide impact of a guest misusing atomic instructions.
   - Support for features new in Intel Granite Rapids CPUs:
     - AVX512-FP16.
 - On Arm:
   - Xen supports guests running SVE/SVE2 instructions. (Tech Preview)
   - Add suport for Firmware Framework for Arm A-profile (FF-A) Mediator (Tech
     Preview)
   - Experimental support for dynamic addition/removal of Xen device tree
     nodes using a device tree overlay binary (.dtbo).
 - Introduce two new hypercalls to map the vCPU runstate and time areas by
   physical rather than linear/virtual addresses.
 - The project has now officially adopted 6 directives and 65 rules of MISRA-C.

### Removed
 - On x86, the "pku" command line option has been removed.  It has never
   behaved precisely as described, and was redundant with the unsupported
   "cpuid=no-pku".  Visibility of PKU to guests should be via its vm.cfg file.
 - xenpvnetboot removed as unable to convert to Python 3.

## [4.17.0](https://xenbits.xenproject.org/gitweb/?p=xen.git;a=shortlog;h=RELEASE-4.17.0) - 2022-12-12

### Changed
 - On x86 "vga=current" can now be used together with GrUB2's gfxpayload setting. Note that
   this requires use of "multiboot2" (and "module2") as the GrUB commands loading Xen.
 - The "gnttab" option now has a new command line sub-option for disabling the
   GNTTABOP_transfer functionality.
 - The x86 MCE command line option info is now updated.

### Added / support upgraded
 - Out-of-tree builds for the hypervisor now supported.
 - __ro_after_init support, for marking data as immutable after boot.
 - The project has officially adopted 4 directives and 24 rules of MISRA-C,
   added MISRA-C checker build integration, and defined how to document
   deviations.
 - IOMMU superpage support on x86, affecting PV guests as well as HVM/PVH ones
   when they don't share page tables with the CPU (HAP / EPT / NPT).
 - Support for VIRT_SSBD and MSR_SPEC_CTRL for HVM guests on AMD.
 - Improved TSC, CPU, and APIC clock frequency calibration on x86.
 - Support for Xen using x86 Control Flow Enforcement technology for its own
   protection. Both Shadow Stacks (ROP protection) and Indirect Branch
   Tracking (COP/JOP protection).
 - Add mwait-idle support for SPR and ADL on x86.
 - Extend security support for hosts to 12 TiB of memory on x86.
 - Add command line option to set cpuid parameters for dom0 at boot time on x86.
 - Improved static configuration options on Arm.
 - cpupools can be specified at boot using device tree on Arm.
 - It is possible to use PV drivers with dom0less guests, allowing statically
   booted dom0less guests with PV devices.
 - On Arm, p2m structures are now allocated out of a pool of memory set aside at
   domain creation.
 - Improved mitigations against Spectre-BHB on Arm.

tools/misc/xenwatchdogd.c
/* SPDX-License-Identifier: GPL-2.0-only */
/*
 * xenwatchdogd.c
 *
 * Watchdog based on Xen hypercall watchdog interface.
 *
 * Copyright 2010 Citrix Ltd
 * Copyright 2024 Leigh Brown <leigh@solinno.co.uk>
 *
 */

#include <err.h>
#include <limits.h>
#include "xenctrl.h"
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <stdio.h>
#include <stdbool.h>
#include <getopt.h>

#define WDOG_MIN_TIMEOUT 2
#define WDOG_MIN_SLEEP 1
#define WDOG_EXIT_TIMEOUT 300

static xc_interface *h;
static volatile bool safeexit = false;
static volatile bool done = false;

static void daemonize(void)
{
    switch (fork()) {
    case -1:
	err(EXIT_FAILURE, "fork");
    case 0:
	break;
    default:
	exit(EXIT_SUCCESS);
    }
    umask(0);
    if (setsid() < 0)
	err(EXIT_FAILURE, "setsid");
    if (chdir("/") < 0)
	err(EXIT_FAILURE, "chdir /");
    if (freopen("/dev/null", "r", stdin) == NULL)
        err(EXIT_FAILURE, "reopen stdin");
    if(freopen("/dev/null", "w", stdout) == NULL)
        err(EXIT_FAILURE, "reopen stdout");
    if(freopen("/dev/null", "w", stderr) == NULL)
        err(EXIT_FAILURE, "reopen stderr");
}

static void catch_exit(int sig)
{
    done = true;
}

static void catch_usr1(int sig)
{
    safeexit = true;
    done = true;
}

static void __attribute__((noreturn)) usage(int exit_code)
{
    FILE *out = exit_code ? stderr : stdout;

    fprintf(out,
	"Usage: xenwatchdog [OPTION]... <timeout> [<sleep>]\n"
	"  -h, --help\t\tDisplay this help text and exit.\n"
	"  -F, --foreground\tRun in foreground.\n"
	"  -x, --safe-exit\tDisable watchdog on orderly exit.\n"
	"\t\t\tNote: default is to set a %d second timeout on exit.\n\n"
	"  timeout\t\tInteger seconds to arm the watchdog each time.\n"
	"\t\t\tNote: minimum timeout is %d seconds.\n\n"
	"  sleep\t\t\tInteger seconds to sleep between arming the watchdog.\n"
	"\t\t\tNote: sleep must be at least %d and less than timeout.\n"
	"\t\t\tIf not specified then set to half the timeout.\n",
	WDOG_EXIT_TIMEOUT, WDOG_MIN_TIMEOUT, WDOG_MIN_SLEEP
	);
    exit(exit_code);
}

static int parse_secs(const char *arg, const char *what)
{
    char *endptr;
    unsigned long val;

    val = strtoul(arg, &endptr, 0);
    if (val > INT_MAX || *endptr)
	errx(EXIT_FAILURE, "invalid %s: '%s'", what, arg);

    return val;
}

int main(int argc, char **argv)
{
    int id;
    int t, s;
    int ret;
    bool daemon = true;

    for ( ;; )
    {
	int option_index = 0, c;
	static const struct option long_options[] =
	{
	    { "help", no_argument, NULL, 'h' },
	    { "foreground", no_argument, NULL, 'F' },
	    { "safe-exit", no_argument, NULL, 'x' },
	    { NULL, 0, NULL, 0 },
	};

	c = getopt_long(argc, argv, "hFxD", long_options, &option_index);
	if (c == -1)
	    break;

	switch (c)
	{
	case 'h':
	    usage(EXIT_SUCCESS);

	case 'F':
	    daemon = false;
	    break;

	case 'x':
	    safeexit = true;
	    break;

	default:
	    usage(EXIT_FAILURE);
	}
    }

    if (argc - optind < 1)
	errx(EXIT_FAILURE, "timeout must be specified");

    if (argc - optind > 2)
	errx(EXIT_FAILURE, "too many arguments");

    t = parse_secs(argv[optind], "timeout");
    if (t < WDOG_MIN_TIMEOUT)
	errx(EXIT_FAILURE, "Error: timeout must be at least %d seconds",
			   WDOG_MIN_TIMEOUT);

    ++optind;
    if (optind < argc) {
	s = parse_secs(argv[optind], "sleep");
	if (s < WDOG_MIN_SLEEP)
	    errx(EXIT_FAILURE, "Error: sleep must be no less than %d",
			       WDOG_MIN_SLEEP);
	if (s >= t)
	    errx(EXIT_FAILURE, "Error: sleep must be less than timeout");
    }
    else
	s = t / 2;

    if (daemon)
	daemonize();

    h = xc_interface_open(NULL, NULL, 0);
    if (h == NULL)
	err(EXIT_FAILURE, "xc_interface_open");

    if (signal(SIGHUP, &catch_exit) == SIG_ERR)
	err(EXIT_FAILURE, "signal");
    if (signal(SIGINT, &catch_exit) == SIG_ERR)
	err(EXIT_FAILURE, "signal");
    if (signal(SIGQUIT, &catch_exit) == SIG_ERR)
	err(EXIT_FAILURE, "signal");
    if (signal(SIGTERM, &catch_exit) == SIG_ERR)
	err(EXIT_FAILURE, "signal");
    if (signal(SIGUSR1, &catch_usr1) == SIG_ERR)
	err(EXIT_FAILURE, "signal");

    id = xc_watchdog(h, 0, t);
    if (id <= 0)
        err(EXIT_FAILURE, "xc_watchdog setup");

    while (!done) {
        sleep(s);
        ret = xc_watchdog(h, id, t);
        if (ret != 0)
            err(EXIT_FAILURE, "xc_watchdog");
    }

    // Zero seconds timeout will disarm the watchdog timer
    xc_watchdog(h, id, safeexit ? 0 : WDOG_EXIT_TIMEOUT);
    return 0;
}
